### PR TITLE
Associated Works for Zenodo Records

### PR DESCRIPTION
**Pull request by tjacovich**
Adds associated works to citation targets in database. Updates current citation targets when a new associated target is added. Adds maintenance task to find associated works for all entries. Includes associated works in forwarded messages to Master.

### CODE REVIEW

**Comment by tjacovich** [comment]
Issue with manually added alternate bibcodes: If there are no alternate bibcodes in the parsed_metadata, parsed_metadata does not have the attribute alternate_bibcode and the generate_modified_metadata fails to set the value and clears alternate_bibcodes from curated_metadata.

**Comment by tjacovich** [comment]
Issue with manually added alternate bibcodes: If there are no alternate bibcodes in the parsed_metadata, parsed_metadata does not have the attribute alternate_bibcode and the generate_modified_metadata fails to set the value and clears alternate_bibcodes from curated_metadata.

**Review by marblestation — APPROVED** [review]
This looks good! I have some suggestions, but I believe it is basically ready to be merged

File: ADSCitationCapture/tests/test_tasks.py [hunk 1]
```diff
@@ -1019,6 +1083,73 @@ def test_process_new_citation_changes_doi_http_error(self):
...
+        citation_change.resolved = False
+        citation_change.status = adsmsg.Status.updated
+        bibcode_id = "2017zndo....248351D"
+        associated_registered_record = self.mock_data[bibcode_id][0]
+        associated_citation_change = adsmsg.CitationChange(content=associated_registered_record['content'],
+                                                       content_type=getattr(adsmsg.CitationChangeContentType, associated_registered_record['content_type'].lower()),
+                                                       status=adsmsg.Status.updated,
+                                                       timestamp=datetime.now()
+                                                       )
+        associated_versions = {"v2.0.0": "2017zndo....248351D", "v3.3.4": "2021zndo...4475376C"}
```
[the remark this answers is not in the document]
Comment (tjacovich) [reply ADSCitationCapture/tests/test_tasks.py:1102]: Added Version

File: ADSCitationCapture/doi.py [hunk 2]
```diff
@@ -198,3 +198,63 @@ def _parse_metadata_zenodo_doi(raw_metadata):
...
+    #check if target is a software version and not the base doi.
+        try:
+            logger.info("{} is version of: {}".format(parsed_metadata['bibcode'], parsed_metadata.get('version_of',None)))
+            #try to recover the base doi for the target
+            raw_metadata = fetch_metadata(base_doi_url, base_datacite_url, parsed_metadata.get('version_of')[0])
+            parsed_all_version = parse_metadata(raw_metadata)
+            if parsed_all_version is not None:
+                logger.debug("Found Associated Versions: {}".format(parsed_all_version.get('versions',None)))
+            #return dois for all versions of the target software and the base doi.
+            versions_json = {'all_doi': parsed_metadata.get('version_of',None)[0], 'versions': parsed_all_version.get('versions',None)}
```
[the remark this answers is not in the document]
Comment (tjacovich) [reply ADSCitationCapture/doi.py:222]: Done

File: README.md [hunk 3]
```diff
@@ -284,9 +284,9 @@ Download the compiled version of montysolr and run it:
 
 ```
 https://github.com/romanchyla/montysolr/releases
-wget -c https://github.com/romanchyla/montysolr/releases/download/v63.1.0.36/montysolr.zip
+wget -c https://github.com/romanchyla/montysolr/releases/download/v77.1.1.4/montysolr.zip
 unzip montysolr.zip
-cd montysolr/
+cd montysolr/ (In Linux, unzip may not create the containing folder and instead unzip the contents into the current directory.)
```
[the remark this answers is not in the document]
Comment (tjacovich) [reply README.md:289]: Updated

File: ADSCitationCapture/tasks.py [hunk 4]
```diff
@@ -201,14 +202,74 @@ def task_process_updated_citation(citation_change, force=False):
...
+            no_self_ref_versions = {key:val for key, val in associated_works.items() if val != citation_target_bibcode} if associated_works else None
             original_citations = db.get_citations_by_bibcode(app, citation_target_bibcode)
             citations = api.get_canonical_bibcodes(app, original_citations)
             logger.debug("Calling 'task_output_results' with '%s'", citation_change)
-            task_output_results.delay(citation_change, parsed_metadata, citations)
+            task_output_results.delay(citation_change, parsed_metadata, citations, db_versions = no_self_ref_versions)
         logger.debug("Calling '_emit_citation_change' with '%s'", citation_change)
         _emit_citation_change(citation_change, parsed_metadata)
 
+def _collect_associated_works(citation_change, parsed_metadata):
```
[the remark this answers is not in the document]
Comment (tjacovich) [reply ADSCitationCapture/tasks.py:215]: docstring explaining what updated function does.

File: ADSCitationCapture/db.py [hunk 5]
```diff
@@ -36,7 +36,7 @@ def store_event(app, data):
             stored = True
     return stored
 
-def store_citation_target(app, citation_change, content_type, raw_metadata, parsed_metadata, status):
+def store_citation_target(app, citation_change, content_type, raw_metadata, parsed_metadata, status, associated=None):
```
[the remark this answers is not in the document]
Comment (tjacovich) [reply ADSCitationCapture/db.py:39]: Homogenizing around no space when in function def or calls, but space when assigning to variable.

**Comment by tjacovich** [comment]
Formatting updates are largely complete, but need to wait to rebase on patched manual curation

**Pull request merged**